switchboard/models.py

```diff
@@ -66,172 +82,129 @@ def create(cls, **kwargs):
         whether a new object was created.
         '''
-        result = cls.c.find_one(kwargs)
-        if not result:
+        instance = cls.get(key)
+        if not instance:
             created = True
-            result = kwargs
-            result.update(defaults)
+            data = dict(key=key)
+            data.update(defaults)
             # Do an upsert here instead of a straight create to avoid a race
             # condition with another instance creating the same record at
             # nearly the same time.
-            cls.update(result, result, upsert=True)
-            result = cls.c.find_one(kwargs)
-            instance = cls(**result)
+            instance = cls.update(data, data, upsert=True)
         else:
             created = False
-            instance = cls(**result)
         return instance, created
 
     @classmethod
-    def find(cls, **kwargs):
-        return [cls(**s) for s in cls.c.find(kwargs)]
-
-    @classmethod
-    def update(cls, spec, document, upsert=False):
+    def update(cls, spec, updates, upsert=False):
         '''
-        Mimics a subset of PyMongo's Collection.update functionality. The spec
-        is used to search for the document to update, document contains the
+        The spec is used to search for the data to update, updates contains the
         values to be updated, and upsert specifies whether to do an insert if
-        the original document is not found.
+        the original data is not found.
         '''
-        previous = cls.get(**spec)
-        cls.pre_save.send(previous)
-        result = cls.c.update(spec, document, upsert=upsert)
-        current = cls.get(**spec)
-        cls.post_save.send(current)
-        return result
+        if 'key' in spec:
+            previous = cls.get(spec['key'])
+        else:
+            previous = None
+        if previous:
+            # Update existing data.
+            current = cls(**previous.__dict__)
+        elif upsert:
+            # Create new data.
+            current = cls(**spec)
+        else:
+            current = None
+        # XXX Should there be any error thrown if this is a noop?
+        if current:
+            current.__dict__.update(updates)
+            current.save()
+        return current
 
     @classmethod
-    def remove(cls, **kwargs):
-        instance = cls.get(**kwargs)
-        cls.pre_delete.send(instance)
-        result = cls.c.remove(kwargs)
-        cls.post_delete.send(instance)
+    def remove(cls, key):
+        key = _key(key)
+        instance = cls.get(key)
+        if instance:
+            cls.pre_delete.send(instance)
+            result = cls.ds.delete(key)
+            cls.post_delete.send(instance)
+        else:
+            # XXX Should there be any error thrown if this is a noop?
+            result = None
         return result
 
     @classmethod
     def all(cls):
-        return [cls(**s) for s in cls.c.find()]
+        if cls._is_redis_datastore():
```

Comment:
Checking specifically for redis here isn't ideal.  It makes me think other datastores could suffer from the same limitations (lack of querying).  Can we feature-detect query support rather than redis specifically?